File: gobites_mobile_application/lib/models/restaurant.dart
import 'package:gobites/models/user.dart';

class Restaurant extends User {
  int rid;
  String restaurantname;
  String ownername;
  String restaurantstyle;
  String address;
  String email;
  String telephoneNo;
  double rating;

  Restaurant(
      {id,
      this.rid,
      username,
      userType,
      this.restaurantname,
      this.ownername,
      this.restaurantstyle,
      this.address,
      this.email,
      this.telephoneNo,
      this.rating})
      : super(id: id, username: username, userType: userType);

  Restaurant.copy(Restaurant from)
      : this(
            id: from.id,
            rid: from.rid,
            username: from.username,
            userType: from.userType,
            restaurantname: from.restaurantname,
            ownername: from.ownername,
            restaurantstyle: from.restaurantstyle,
            address: from.address,
            email: from.email,
            telephoneNo: from.telephoneNo,
            rating: from.rating);

  Restaurant.fromJson(Map<String, dynamic> json)
      : this(
            id: json['id'],
            rid: json['RID'],
            username: json['username'],
            userType: json['usertype'],
            restaurantname: json['restaurantname'],
            ownername: json['ownername'],
            restaurantstyle: json['restaurantstyle'],
            address: json['address'],
            email: json['email'],
            telephoneNo: json['telephoneNo'],
            rating: json['rating'] == null ? -1 : json['rating'] + .0);

  Map<String, dynamic> toJson() => {
        'id': id,
        'username': username,
        'usertype': userType,
        'restaurantname': restaurantname,
        'ownername': ownername,
        'restaurantstyle': restaurantstyle,
        'address': address,
        'email': email,
        'telephoneNo': telephoneNo,
      };
}

File: gobites_server/models/restaurant.js
const db = require("../db");

class Restaurant {
  constructor() {
    if (this.instance) return this.instance;
    Restaurant.instance = this;
  }

  async get(id) {
    if (id) {
      const rows = await db.query(
        `SELECT * FROM restaurant 
      INNER JOIN user ON user.fk_rid=restaurant.RID 
      WHERE user.UID = ?`,
        [id]
      );
      return this.rowToArray(rows[0]);
    }
    const rows = await db.query("SELECT * FROM restaurant");
    return this.rowToArray(rows);
  }

  async getByOid(oid) {
      const rows = await db.query(
              `SELECT restaurant.restaurantname, restaurant.RID
            FROM orderitem
            JOIN menuitem ON menuitem.MID = orderitem.fk_mid
            JOIN restaurant ON restaurant.RID = menuitem.fk_rid
            WHERE orderitem.fk_oid=?
            GROUP BY restaurant.restaurantname, restaurant.RID
            `,
        [oid]
      );
      return this.rowToArray(rows[0]);
  }

  async put(
    id,
    { restaurantname, ownername, address, restaurantstyle, email, telephoneNo }
  ) {
    console.log(ownername);
    const rows = await db.query(
      `UPDATE restaurant
      SET restaurantname = ?, ownername = ?, address = ?,restaurantstyle= ?, email = ?, telephoneNo = ?
      WHERE RID = ?`,
      [
        restaurantname,
        ownername,
        address,
        restaurantstyle,
        email,
        telephoneNo,
        id,
      ]
    );
    return this.rowToArray(rows);
  }

  rowToArray(sqlRows) {
    if (!sqlRows) return null;
    return JSON.parse(JSON.stringify(sqlRows));
  }
}

module.exports = new Restaurant();
